barretenberg/sol/src/honk/utils.sol

```diff
@@ -54,6 +54,135 @@ function negateInplace(Honk.G1Point memory point) pure returns (Honk.G1Point mem
     return point;
 }
 
+function convertPairingPointsToG1(Fr[PAIRING_POINTS_SIZE] memory pairingPoints)
+    pure
+    returns (Honk.G1Point memory lhs, Honk.G1Point memory rhs)
+{
+    uint256 lhsX = Fr.unwrap(pairingPoints[0]);
+    lhsX |= Fr.unwrap(pairingPoints[1]) << 68;
+    lhsX |= Fr.unwrap(pairingPoints[2]) << 136;
+    lhsX |= Fr.unwrap(pairingPoints[3]) << 204;
+    lhs.x = lhsX;
+
+    uint256 lhsY = Fr.unwrap(pairingPoints[4]);
+    lhsY |= Fr.unwrap(pairingPoints[5]) << 68;
+    lhsY |= Fr.unwrap(pairingPoints[6]) << 136;
+    lhsY |= Fr.unwrap(pairingPoints[7]) << 204;
+    lhs.y = lhsY;
+
+    uint256 rhsX = Fr.unwrap(pairingPoints[8]);
+    rhsX |= Fr.unwrap(pairingPoints[9]) << 68;
+    rhsX |= Fr.unwrap(pairingPoints[10]) << 136;
+    rhsX |= Fr.unwrap(pairingPoints[11]) << 204;
+    rhs.x = rhsX;
+
+    uint256 rhsY = Fr.unwrap(pairingPoints[12]);
+    rhsY |= Fr.unwrap(pairingPoints[13]) << 68;
+    rhsY |= Fr.unwrap(pairingPoints[14]) << 136;
+    rhsY |= Fr.unwrap(pairingPoints[15]) << 204;
+    rhs.y = rhsY;
+}
+
+function convertG1ToPairingPoints(Honk.G1Point memory lhs, Honk.G1Point memory rhs)
+    pure
+    returns (Fr[PAIRING_POINTS_SIZE] memory pairingPoints)
+{
+    // lhs.x
+    pairingPoints[0] = Fr.wrap(uint64(lhs.x));
+    pairingPoints[1] = Fr.wrap(uint64(lhs.x >> 68));
+    pairingPoints[2] = Fr.wrap(uint64(lhs.x >> 136));
+    pairingPoints[3] = Fr.wrap(uint64(lhs.x >> 204));
+    // lhs.y
+    pairingPoints[4] = Fr.wrap(uint64(lhs.y));
+    pairingPoints[5] = Fr.wrap(uint64(lhs.y >> 68));
+    pairingPoints[6] = Fr.wrap(uint64(lhs.y >> 136));
+    pairingPoints[7] = Fr.wrap(uint64(lhs.y >> 204));
+    // rhs.x
+    pairingPoints[8] = Fr.wrap(uint64(rhs.x));
+    pairingPoints[9] = Fr.wrap(uint64(rhs.x >> 68));
+    pairingPoints[10] = Fr.wrap(uint64(rhs.x >> 136));
+    pairingPoints[11] = Fr.wrap(uint64(rhs.x >> 204));
+    // rhs.y
+    pairingPoints[12] = Fr.wrap(uint64(rhs.y));
+    pairingPoints[13] = Fr.wrap(uint64(rhs.y >> 68));
+    pairingPoints[14] = Fr.wrap(uint64(rhs.y >> 136));
+    pairingPoints[15] = Fr.wrap(uint64(rhs.y >> 204));
+}
+
+function generateRecursionSeparator(
+    Fr[PAIRING_POINTS_SIZE] memory proofPairingPoints,
+    Honk.G1Point memory accLhs,
+    Honk.G1Point memory accRhs
+) pure returns (Fr recursionSeparator) {
+    // hash the proof aggregated X
+    // hash the proof aggregated Y
+    // hash the accum X
+    // hash the accum Y
+
+    uint256[PAIRING_POINTS_SIZE * 2] memory recursionSeparatorElements;
```

Comment:
since the source of this randomness doesnt matter too much, ill not copy what the circuit does directly, and ill hash the pairing points compressed rather than large in an array like this